### ISSUE TITLE
Update labels for n>N turns and variants (reversing)

### ISSUE THREAD

**Issue by ManDay** [8VIM/8VIM#241]
I think it is a good choice to **not** indicate "live" information on the sector-cross while writing a character (i.e. not display one character in each sector after a turn has begun), because this would be too dynamic and confusing, while the user could no longer peek ahead for their next input.

However, I think an update of the labels would be appropriate when the turns exceed *N = 4*, the number of sectors (= applying *Shift*) or when reversing direction to access nested inputs (= diacritics/variants).
 - Labels would change to the Shift-ed labels (generically a set indexed by *n*, the number of turns *div N*, the number of sectors) when exceeding *n > N*
 - Labels would change when the direction has been reversed

**Comment by davidoskky** [comment]
Displaying live information might be useful for beginners in my opinion.
Of course this should be a setting that can be disabled.
I do agree that showing it in the sectors might be confusing.
Showing the currently selected letter in the circle may be a better option.
In this way you move to the circle to "accept" the currently displayed character.

This doesn't hinder the visibility of the sectors and shouldn't create too many problems, unless you're typing very fast and keep seeing changing letters in the circle, which might be confusing.

**Comment by doak** [comment]
I second the idea because there will be always foreign characters you don't remember because you write them four times a year.
(I assume a (custom) keyboard with lot of characters.)
Without a live preview in such cases, you won't be able to use such a keyboard at all.

Hence the idea to switch the labels when effectively switching layers is a good one, though you may use a delay before actually doing it. This way it does not disturb, but as soon as the user make a pause to think about the location, it will be ready on his/her finger tips :P

**Comment by rickyson96** [comment]
Hi! I am a new user of this keyboard and just started to learn how to use the feature.
One thing that I'm missing is what is discussed here, a live preview feature, to help ease me on learning the keyboard :joy:

I am thinking, instead of updating label or show it in the circle, maybe we can use a popup (similar to what traditional keyboard pops up when typed). And in the popup, we can highlight currently selected characters and display other characters when we have crossed one of the sector

Something like this
<img width="364" alt="image" src="https://user-images.githubusercontent.com/43179681/156794374-8d52b131-f477-406e-b82e-c878bd6ef404.png"> <img width="363" alt="image" src="https://user-images.githubusercontent.com/43179681/156794459-8f0bc445-679a-467e-8d4b-107fbaaffef4.png">

We can also shows the popup when user starts to enter one of the area (when choosing which sector they want to enter)
<img width="355" alt="image" src="https://user-images.githubusercontent.com/43179681/156795412-b579e238-d680-4f69-845f-6a7dcb4e14d3.png">

**Comment by flide** [comment]
@rickyson96 I love you idea.

The biggest problem is setting up a ... what are we calling this? 'live-preview'? 'letter-preview'? 'letter-assist'?...
anyways, the biggest problem with setting this up is the fact that all the letters are completely configurable, so the app has no idea what the letter is at any given place ahead of time, there might not be any, I don't know how to explain this...

The one solution for this would be set up a different format for the layout files and allow specific number of entries in it, all mandatory. No other gestures would be allowed in the files, if this is achieved, then and only then would I be able to add 'letter-preview' (yea I picked one) feature with the mental peace that the app won't crash.

**Comment by rickyson96** [comment]
I'm thinking of setting up the 'letter-preview' feature the way you've set up the labels on the main keyboard (depends on the character list pulled from the configuration selected by the user) and limiting it to only the gestures that start from the inner circle if that makes sense.
The thing is that I'm neither familiar with the codebase, nor with android/IME development, so I might be misleading with the comment above.

> ... I be able to add 'letter-preview' (yea I picked one) ...

I'm thinking of something like `popup-assistant` or something that indicates that it's a popup, but I'm happy with the `letter-preview` too.

Anyway, thanks for the great work and app! Really enjoyed using it (though I'm still learning to use it).

**Comment by Paperweb** [comment]
@rickyson96 Why not use the circle to display the current glyph, as @davidoskky suggests ?
I feel that an extra popup adds a layer of stuff happening on the screen, when we already have the info written in the sectors.

So instead we could use the already-there info, and highlight it according to the gesture.

I didn't think of it at first, but i agree with @ManDay that a live update that fades out too much the other sectors and glyphs could be confusing to prepare the next inputs.

So to summarize, to me, the solution should be one where you can : 
- see the current glyph "in cache", for example inside the circle
- highlight the current glyph in the sector to know "where" you are in the sequence (bold, color, others lightly fade)
- dynamically change labels when n>N and when going back one sector to produce variants (see also #65)

I genuinely believe this is one of the most needed features of the app to facilitate learning and retain users !

**Comment by Paperweb** [comment]
> The biggest problem with setting this up is the fact that all the letters are completely configurable, so the app has no idea what the letter is at any given place ahead of time, there might not be any, I don't know how to explain this

Oh i see @flide , do you mean that the app is not populating the display as you go, but instead populates it once, and then reads from the layout file for each input?

**Comment by flide** [comment]
yes

**Comment by rickyson96** [comment]
@Paperweb I am suggesting popup because it feels like my thumb is constantly blocking the view to the circle and sectors while typing, while with popup, we can place it at the top so that we can view it clearly.

**Comment by flide** [comment]
does #271 fix this? can this be closed?

**Comment by ManDay** [comment]
I think what's still missing is some indicator for diacritics/reversal. Otherwise looks fine!

**Comment by ygrek** [comment]
> my thumb is constantly blocking the view to the circle and sectors while typing

strongly agree with this, current highlight is very limited use because i (as a very beginner) need to stop typing and awkwardly look under the thumb to confirm what is the current letter. Live preview in the top sector would be a huge win (even just the current letter, can think of showing other characters later).

PS https://github.com/8VIM/8VIM/issues/420
